runtime-manager/src/managers/execution_engine_manager.rs

```diff
@@ -172,33 +177,52 @@ fn parse_incoming_buffer(
     let mut incoming_buffer_hash = INCOMING_BUFFER_HASH.lock()?;
 
     // First, make sure there is an entry in the hash for the TLS session.
-    if incoming_buffer_hash.get(&tls_session_id).is_none() {
-        incoming_buffer_hash.insert(tls_session_id, Vec::new());
+    // If not, this means we are receiving the first chunk of the protocol
+    // buffer.
+    let mut input = if incoming_buffer_hash.get(&tls_session_id).is_none() {
+        // Extract the protocol buffer's total length as u64
+        let input_data = input.split_off(8);
+        let mut expected_length: [u8; 8] = [0; 8];
+        expected_length.copy_from_slice(input.as_slice());
+        let expected_length = u64::from_be_bytes(expected_length);
+
+        // Insert the expected length in the hash table
+        incoming_buffer_hash.insert(tls_session_id, (expected_length, Vec::new()));
+
+        input_data
     }
+    else {
+        input
+    };
 
     // This should not panic, given the above.  If it does, something is wrong.
-    let incoming_buffer = incoming_buffer_hash.get_mut(&tls_session_id).ok_or(
+    let (expected_length, incoming_buffer) = incoming_buffer_hash.get_mut(&tls_session_id).ok_or(
         RuntimeManagerError::UnavailableIncomeBufferError(tls_session_id as u64),
     )?;
 
     incoming_buffer.append(&mut input);
 
-    // NB: `parse_from_bytes()` returning failure is interpreted as meaning the
-    // full protocol buffer has not yet been received. So we return `Ok(None)`
-    // with the hope that eventually we will receive all of it, and then
-    // `parse_from_bytes()` will return `Ok(parsed)`. In a well-behaving system,
-    // this is reasonable.  In a poorly-behaved system (under attack, clients
-    // just getting confused) it is not reasonable, and can eventually result in
-    // "Out of Memory" or Garbage out.
+    // We return `Ok(None)` as long as the full protocol buffer has not yet been
+    // received. Once received, `parse_from_bytes()` parses it and returns
+    // `Ok(parsed)`.
+    // In a well-behaving system, this is reasonable.  In a poorly-behaved
+    // system (under attack, clients just getting confused) it is not
+    // reasonable, and can eventually result in "Out of Memory" or Garbage out.
```

Comment:
Surely we should never allow "garbage out" under any circumstances?  What can be done to stop this?

Reply:
Agreed. We could enforce a maximum amount of memory per session/enclave as part of the policy, and even check beforehand if the file to be sent by the client can fit in the runtime manager

Reply:
Ah OK, I misunderstood the behaviour of the `split_off` function: I thought it returned the first 8 bytes and modified the original vector in-place to point-to the resulting tail.

Reply:
I'd really like a function like that but couldn't find it

Reply:
I think adding a new policy field to specify the amount of memory available per session should be part of another PR, as it was already a problem before. I'm opening an issue for that

Reply:
OK, that sounds like a good plan.